In [162]:
# The below lines of code import the pandas library and the display function.
import pandas as pd 
from IPython.display import display

# The below lines of code read in the train_df and test_df CSV files.
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# The below lines of code display the first five rows of the train data frame.
print('First five rows of train data frame:')
display(train.head())


print()
print()
# The below lines of code display the first five rows of the test data frame.
print('First five rows of test data frame:')
display(test.head())

# The below lines of code delete the previous index column in both the train and test data frames. 
del train['Unnamed: 0']
del test['Unnamed: 0']


print()
print()
# The below lines of code display the first five rows of the train data frame.
print('First five rows of train data frame after removing unnecesary column:')
display(train.head())

print()
print()
# The below lines of code display the first five rows of the test data frame.
print('First five rows of test data frame after removing unnecesary column:')
display(test.head())

First five rows of train data frame:


Unnamed: 0           City  Zip or postal code     Price  Beds  Baths  \
0        1048         Olathe             66062.0  350000.0   3.0    3.0   
1         406        Mission             66202.0  297000.0   2.0    2.5   
2        1027         Olathe             66062.0  279950.0   3.0    2.0   
3         486  Overland Park             66204.0  299000.0   3.0    1.0   
4        1485         Lenexa             66215.0  531000.0   4.0    3.5   

         Location  Square feet  Lot size  Year built  Hoa/month  
0      Quail Park       1876.0   11761.0      1995.0       13.0  
1   DILLE HEIGHTS       1559.0   15001.0      1990.0        0.0  
2  Arrowhead East       1344.0    7405.0      1982.0        0.0  
3   Rayven Plains       1404.0   11353.0      1950.0        0.0  
4      Oak Valley       3542.0   12553.0      1994.0       20.0



First five rows of test data frame:


Unnamed: 0           City  Zip or postal code     Price  Beds  Baths  \
0         917         Olathe             66061.0  749500.0   6.0    4.0   
1         134  Overland Park             66224.0  939000.0   4.0    4.5   
2         390        Shawnee             66203.0  265000.0   3.0    2.0   
3         790         Olathe             66061.0  308000.0   3.0    2.0   
4        1498         Lenexa             66215.0  345000.0   3.0    2.5   

                     Location  Square feet  Lot size  Year built  Hoa/month  
0  Cedar Creek - Valley Ridge       2884.0    9587.0      2023.0      156.0  
1                  WatersEdge       3880.0   11944.0      2016.0      131.0  
2              Tomahawk Hills       1136.0    9877.0      1962.0        0.0  
3               Rose Addition       1808.0   12207.0      1964.0        0.0  
4             Century Estates       1766.0   14249.0      1972.0        0.0



First five rows of train data frame after removing unnecesary column:


City  Zip or postal code     Price  Beds  Baths        Location  \
0         Olathe             66062.0  350000.0   3.0    3.0      Quail Park   
1        Mission             66202.0  297000.0   2.0    2.5   DILLE HEIGHTS   
2         Olathe             66062.0  279950.0   3.0    2.0  Arrowhead East   
3  Overland Park             66204.0  299000.0   3.0    1.0   Rayven Plains   
4         Lenexa             66215.0  531000.0   4.0    3.5      Oak Valley   

   Square feet  Lot size  Year built  Hoa/month  
0       1876.0   11761.0      1995.0       13.0  
1       1559.0   15001.0      1990.0        0.0  
2       1344.0    7405.0      1982.0        0.0  
3       1404.0   11353.0      1950.0        0.0  
4       3542.0   12553.0      1994.0       20.0



First five rows of test data frame after removing unnecesary column:


City  Zip or postal code     Price  Beds  Baths  \
0         Olathe             66061.0  749500.0   6.0    4.0   
1  Overland Park             66224.0  939000.0   4.0    4.5   
2        Shawnee             66203.0  265000.0   3.0    2.0   
3         Olathe             66061.0  308000.0   3.0    2.0   
4         Lenexa             66215.0  345000.0   3.0    2.5   

                     Location  Square feet  Lot size  Year built  Hoa/month  
0  Cedar Creek - Valley Ridge       2884.0    9587.0      2023.0      156.0  
1                  WatersEdge       3880.0   11944.0      2016.0      131.0  
2              Tomahawk Hills       1136.0    9877.0      1962.0        0.0  
3               Rose Addition       1808.0   12207.0      1964.0        0.0  
4             Century Estates       1766.0   14249.0      1972.0        0.0

In [ ]:
# Importing transformation functions and libraries
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import FunctionTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
import numpy as np
from sklearn.preprocessing import PowerTransformer



# The below lines of code assign the Price column from the train and test data frames as the target variables.
train_target = train['Price']
test_target = test['Price']

# The below lines of code delete the Price and Location column from the train and test data frames. 
del train['Price']
del test['Price']
del train['Location']
del test['Location']

# Creating list to hold names of numerical and categorical features
cat_feat = ['City','Zip or postal code']
num_feat = ['Beds','Baths', 'Square feet', 'Lot size', 'Year built', 'Hoa/month']

# Creating pipe1, which will logarithmically transform the features and then scale them.
pipe1 = Pipeline([
    ('log', FunctionTransformer(func=np.log1p)),
    ('scaler', StandardScaler()),
    ])
# Creating a column transform that will send the numerical features to pipe1 and then one-hot encode the categorical features (while dropping the first column).
big_pipeline = ColumnTransformer([
    ('num', pipe1, num_feat),
    ('cat', OneHotEncoder(drop='first'), cat_feat)
])

# Fitting big pipeline with train data, transforming train data using big_pipeline, and transforming testing data using big_pipeline.
train_og = big_pipeline.fit_transform(train)
test_og = big_pipeline.transform(test)

# Creating pipe2, which will yeo-johnson transform the features and then scale them.
pipe2 = Pipeline([
    ('yeo-john', PowerTransformer(method='yeo-johnson', standardize=False)),
    ('scaler', StandardScaler()),
    ])
# Creating a column transformer that will send the numerical features to pipe2 and then one-hot encode the categorical features (while dropping the first column and not outputting as a sparse matrix).
big_pipeline2 = ColumnTransformer([
    ('num', pipe2, num_feat),
    ('cat', OneHotEncoder(drop='first', sparse_output=False), cat_feat)
])

# Fitting big_pipeline2 with train data, transform train data using big_pipeline2, and transforming test data using big_pipeline2.
train_2 = big_pipeline2.fit_transform(train)
test_2 = big_pipeline2.transform(test)


# The below line of code creates a logarithmically transformed target variable.
train_target_log = np.log(train_target)


In [ ]:
# Loading in regressor model, scoring function, and search function to determine best combination of parameters.
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_absolute_error

### Creating first gradient boosting regressor model

In [ ]:
# Fitting a gradient boosting regressor with data trained by big_pipeline1, and testing said model on test data transformed by big_pipeline1.
gbr = GradientBoostingRegressor(random_state=22)
gbr.fit(train_og, train_target_log)
pred_target_log = gbr.predict(train_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target),2)
print('MAE on training data with original pipeline:', mae)
pred_target_log = gbr.predict(test_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target),2)
print('MAE on testing data with original pipeline', mae)


print()
# Fitting a gradient boosting regressor with data trained by big_pipeline2, and testing said model on test data transformed by big_pipeline2.
gbr = GradientBoostingRegressor(random_state=22)
gbr.fit(train_2, train_target_log)
pred_target_log = gbr.predict(train_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target),2)
print('MAE on training data with second pipeline:', mae)
pred_target_log = gbr.predict(test_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target),2)
print('MAE on testing data with second pipeline', mae)

MAE on training data with original pipeline: 49173.64
MAE on testing data with original pipeline 69912.08

MAE on training data with second pipeline: 49173.64
MAE on testing data with second pipeline 69776.05


 The Gradient Boosting Regressor model fit with predictors from the original pipeline has a training MAE of 49173.64 and a testing MAE of 69912.64. Considering that the testing MAE is fairly low, I am surprised that the difference between the training and testing MAE is only around 20,000. In previous notebooks, it seems like whenever this difference decreases, the testing MAE would skyrocket.

The second model contains the same training MAE but slightly improved the testing MAE. 

In [170]:
gbr.get_params()

{'alpha': 0.9,
 'ccp_alpha': 0.0,
 'criterion': 'friedman_mse',
 'init': None,
 'learning_rate': 0.1,
 'loss': 'squared_error',
 'max_depth': 3,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_iter_no_change': None,
 'random_state': 22,
 'subsample': 1.0,
 'tol': 0.0001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

### Creating second Gradient Boosting Regressor models

The goal here is that hopefully by reducing the number of estimators, the model won't have as many chances to get better at knowing the train data as well, and will be able to generalize to unseen data better.

In [ ]:
gbr = GradientBoostingRegressor(random_state=22, n_estimators=75)
# Fitting the gbr with train data transformed by big_pipeline1 and testing gbr with test data transformed by big_pipeline1.
gbr.fit(train_og, train_target_log)
pred_target_log = gbr.predict(train_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('MAE on training data with original pipeline:', mae)
pred_target_log = gbr.predict(test_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE on testing data with original pipeline', mae)

print()
# Fitting the gbr with train data transformed by big_pipeline2 and testing gbr with test data transformed by big_pipeline2.
gbr.fit(train_2, train_target_log)
pred_target_log = gbr.predict(train_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('MAE on training data with second pipeline:', mae)
pred_target_log = gbr.predict(test_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE on testing data with second pipeline', mae)

MAE on training data with original pipeline: 52499.0
MAE on testing data with original pipeline 71541.0

MAE on training data with second pipeline: 52499.0
MAE on testing data with second pipeline 71450.0


Both train and test data MAE increased, indicating the reducing the number of estimators didn't have the impact I was hoping for.

### Attempting to use Random Search CV to find a model that generalizes better to unseen data.

In [ ]:
# The following lines of code create a dictionary with different hyperparameter values.
param_grid = {
    'n_estimators': np.arange(1,200,5),
    'max_depth': np.arange(5,100,5),
    'min_samples_split': np.arange(2,20,1),
    'learning_rate': np.arange(.01,1,.01),
    'subsample': np.arange(.01,1,.01),
}


# The below line of code specifies an initial Gradient Boosting Regressor.
gbr = GradientBoostingRegressor(random_state=42)

# The following initializes a RandomizedSearchCV model with the Gradient Boosting Regressor, using param_grid for hyperparameters, neg_mean_absolute_error as the scoring metric, 4-fold cross-validation, and a random state for reproducibility.
rscv = RandomizedSearchCV(gbr, param_distributions=param_grid, scoring='neg_mean_absolute_error', cv=4, return_train_score=True, random_state=22)

# Fitting the randomizedsearchCV with data transformed by big_pipeline2.
rscv.fit(train_2, train_target_log)

RandomizedSearchCV(cv=4, estimator=GradientBoostingRegressor(random_state=42),
                   param_distributions={'learning_rate': array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 , 0.11,
       0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 , 0.21, 0.22,
       0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 , 0.31, 0.32, 0.33,
       0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 , 0.41, 0.42, 0.43, 0.44,
       0.45, 0.46, 0.4...
       0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 , 0.41, 0.42, 0.43, 0.44,
       0.45, 0.46, 0.47, 0.48, 0.49, 0.5 , 0.51, 0.52, 0.53, 0.54, 0.55,
       0.56, 0.57, 0.58, 0.59, 0.6 , 0.61, 0.62, 0.63, 0.64, 0.65, 0.66,
       0.67, 0.68, 0.69, 0.7 , 0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77,
       0.78, 0.79, 0.8 , 0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88,
       0.89, 0.9 , 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99])},
                   random_state=22, return_train_score=True,
                   scoring='neg_mean_absolute_error')

In [ ]:
rs = rscv.best_estimator_
# Fitting the best estimator from the randomized search function with train data transformed by big_pipeline1 and testing said estimator on test data transformed by big_pipeline1.
rs.fit(train_og, train_target_log)
pred_target_log = rs.predict(train_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('MAE on training data with original pipeline:', mae)
pred_target_log = rs.predict(test_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE on testing data with original pipeline', mae)

print()
# Fitting the best estimator from the randomized search function with train data transformed by big_pipeline2 and testing said estimator on test data transformed by big_pipeline2.
rs.fit(train_2, train_target_log)
pred_target_log = rs.predict(train_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('MAE on training data with second pipeline:', mae)
pred_target_log = rs.predict(test_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE on testing data with second pipeline', mae)

MAE on training data with original pipeline: 43019.0
MAE on testing data with original pipeline 73003.0

MAE on training data with second pipeline: 44263.0
MAE on testing data with second pipeline 72734.0


From the above results, there are still clear signs of innaccuraies and overfitting. To hopefully address this, I plan to use the RandomizedSearchCV function with a custom score. Currently, I've been using either train_target (normal values) or train_target_log (logarithmically transformed value) as what I am feeding into the RandomizedSearchCV function as the target.

Both these will lead to issues.

1. When using train_target, the default scoring will work perfectly fine in terms of scale. However, I have been assuming that when I retrain the model, that the transformed target variables will allow the model to perform at its best. 

2. When using the train_target_log, the scoring is based on logarithmically scaled values, which doesn't align with how I plan to evaluate the models. Instead, I need scoring models that can take the logarithmically transformed true target values and transform them properly, and then transformed the predicted target values, and then I will be able to calculate a suitable mean absolute error.

### Utilizing a custom scorer in the Randomized Search CV function

In [ ]:
# Importing the make_scorer function
from sklearn.metrics import make_scorer

# Creating a dictionary for potential parameter options.
param_grid = {
    'n_estimators': np.arange(1,400,5),
    'max_depth': np.arange(1,100,5),
    'min_samples_split': np.arange(2,20,2),
    'learning_rate': np.arange(.05,.35,.03),
    'subsample': np.arange(.1,1,.1),
}

# The below code is property of Gemini. I have had issues before with creating custom scoring functions, so I apologize if anyone reading this notebook does't like that.
def custom_mae(y_true, y_pred):
    y_true_original = np.exp(y_true)
    y_pred_original = np.exp(y_pred)
    return mean_absolute_error(y_true_original, y_pred_original)
custom_scorer = make_scorer(custom_mae, greater_is_better=False) 


# The below line of code specifies an initial Gradient Boosting Regressor.
gbr = GradientBoostingRegressor(random_state=22)

# The following initializes a RandomizedSearchCV model with the Gradient Boosting Regressor, using param_grid for hyperparameters, custom_score as the scoring metric, 5-fold cross-validation, and a random state for reproducibility.
rscv = RandomizedSearchCV(gbr, param_distributions=param_grid, scoring=custom_scorer, cv=5, return_train_score=True, random_state=22)

rscv.fit(train_2, train_target_log)

RandomizedSearchCV(cv=5, estimator=GradientBoostingRegressor(random_state=22),
                   param_distributions={'learning_rate': array([0.05, 0.08, 0.11, 0.14, 0.17, 0.2 , 0.23, 0.26, 0.29, 0.32]),
                                        'max_depth': array([ 1,  6, 11, 16, 21, 26, 31, 36, 41, 46, 51, 56, 61, 66, 71, 76, 81,
       86, 91, 96]),
                                        'min_samples_split': array([ 2,  4,  6,  8, 10, 12, 14, 16, 18]),
                                        'n_estimators': array([  1,   6,  11,  16,  21,  26,  31,  36,  41,...
       131, 136, 141, 146, 151, 156, 161, 166, 171, 176, 181, 186, 191,
       196, 201, 206, 211, 216, 221, 226, 231, 236, 241, 246, 251, 256,
       261, 266, 271, 276, 281, 286, 291, 296, 301, 306, 311, 316, 321,
       326, 331, 336, 341, 346, 351, 356, 361, 366, 371, 376, 381, 386,
       391, 396]),
                                        'subsample': array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])},
                   random_state=22, return_train_score=True,
                   scoring=make_scorer(custom_mae, greater_is_better=False))

Although not all the above code was me, I am still excited that a custom_scoring function was finally made that would work.

In [175]:
rscv.best_estimator_

GradientBoostingRegressor(learning_rate=0.16999999999999998, max_depth=21,
                          min_samples_split=14, n_estimators=321,
                          random_state=22, subsample=0.5)

Looking at the parameters, there does seem to be some clear changes between the default values and the model that was determined to result in the lowest MAE for the training data.

In [ ]:
model = rscv.best_estimator_
# Fitting the best estimator with train data transformed by big_pipeline1 and testing said model on test data transformed by big_pipeline1.
model.fit(train_og, train_target_log)
pred_target_log = model.predict(train_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('MAE on training data with original pipeline:', mae)
pred_target_log = model.predict(test_og)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE on testing data with original pipeline', mae)

print()
# Fitting the best estimator with train data transformed by big_pipeline2 and testing said model on test data transformed by big_pipeline2.
model.fit(train_2, train_target_log)
pred_target_log = model.predict(train_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('MAE on training data with second pipeline:', mae)
pred_target_log = model.predict(test_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE on testing data with second pipeline', mae)


MAE on training data with original pipeline: 3225.0
MAE on testing data with original pipeline 70720.0

MAE on training data with second pipeline: 3241.0
MAE on testing data with second pipeline 70038.0


Both the training MAEs are around 3220, while testing MAEs are around 70000. This difference is perplexing because the RandomizedSearchCV function splits the data into five sections for cross-validation, which should allow the model to generalize well to unseen data by testing it on each held-out section. However, the model performs exceptionally well on the training data but fails to generalize to the testing data. 

One possible reason for this simply could be that the RandomizedSearchCV function was given too many possibilities in terms of parameter combinations.  

### Limiting the number of parameter options in the Randomized Search CV function.

In [238]:
# The following lines of code create a dictionary with different hyperparameter values.
param_grid = {
    'n_estimators': np.arange(1,200,1),
    'max_depth': np.arange(1,10,1),
    'min_samples_split': np.arange(2,10,2),
    'learning_rate': np.arange(.05,1,.25),
    'subsample': np.arange(.1,1,.2),
}

# The below code is property of Gemini. I have had issues before with creating custom scoring functions, so I apologize if anyone reading this notebook does't like that.
def custom_mae(y_true, y_pred):
    y_true_original = np.exp(y_true)
    y_pred_original = np.exp(y_pred)
    return mean_absolute_error(y_true_original, y_pred_original)
custom_scorer = make_scorer(custom_mae, greater_is_better=False) 


# The below line of code specifies an initial Gradient Boosting Regressor.
gbr = GradientBoostingRegressor(random_state=22)

# The following initializes a RandomizedSearchCV model with the Gradient Boosting Regressor, using param_grid for hyperparameters, custom_scorer as the scoring metric, 5-fold cross-validation, and a random state for reproducibility.
rscv = RandomizedSearchCV(gbr, param_distributions=param_grid, scoring=custom_scorer, cv=5, return_train_score=True, random_state=22)

# Fitting rscv with data transformed by big_pipeline2.
rscv.fit(train_2, train_target_log)


model = rscv.best_estimator_
# Fitting the best estimator with train data transformed by big_pipeline2 and testing said model on test data transformed by big_pipeline2.
model.fit(train_2, train_target_log)
pred_target_log = model.predict(train_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('MAE on training data with second pipeline:', mae)
pred_target_log = model.predict(test_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE on testing data with second pipeline', mae)

MAE on training data with second pipeline: 58724.0
MAE on testing data with second pipeline 73305.0


Well, the adjustments we made did lead to the model generalizing better than previously, but the testing MAE is worse than it was earlier, so I don't think the adjustments made are the rescue to this problem.

While it has been easier to just feed the parameters into a function, it may be not working as I hoped unfortunately. 

### Seeing if removing less important features will produce a better regressor.

In [ ]:
# Extracting the feature names from the one-hot encoded columns, combining them with the numerical features, and creating data frames with the data transformed by big_pipeline2.
cat_encoder = big_pipeline2.named_transformers_['cat']
cat_one_hot_attribs = cat_encoder.get_feature_names_out(cat_feat)
attributes = num_feat + list(cat_one_hot_attribs)
train_2 = pd.DataFrame(train_2, columns=attributes)
test_2 = pd.DataFrame(test_2, columns=attributes)

# Creating data frame with feature name and feature importance.
f_impt = pd.DataFrame()
f_impt['features'] = attributes
f_impt['importance'] = model.feature_importances_

# Displaying the feature importance data frame.
display(f_impt.sort_values(by='importance', ascending=False))

# Filtering to only include features with importance greater than 0 and creating a list of those features.
imp = f_impt[f_impt['importance'] > 0]
impt_feat = imp['features'].values.tolist()

features  importance
2                  Square feet    0.457026
3                     Lot size    0.104781
4                   Year built    0.098736
1                        Baths    0.085818
5                    Hoa/month    0.075080
32  Zip or postal code_66208.0    0.037357
0                         Beds    0.016583
14        City_Prairie Village    0.014690
11          City_Mission Hills    0.013400
22  Zip or postal code_66061.0    0.010628
29  Zip or postal code_66205.0    0.010300
36  Zip or postal code_66212.0    0.006433
12                 City_Olathe    0.005811
44  Zip or postal code_66221.0    0.005384
46  Zip or postal code_66224.0    0.005019
31  Zip or postal code_66207.0    0.005009
13          City_Overland Park    0.004917
48  Zip or postal code_66226.0    0.003949
33  Zip or postal code_66209.0    0.003824
7                 City_Leawood    0.003406
18               City_Westwood    0.003180
45  Zip or postal code_66223.0    0.002596
9                 City_Merriam    0.002564
23  Zip or postal code_66062.0    0.002415
16                City_Shawnee    0.002335
37  Zip or postal code_66213.0    0.002232
27  Zip or postal code_66203.0    0.002204
15           City_Roeland Park    0.002054
28  Zip or postal code_66204.0    0.001813
34  Zip or postal code_66210.0    0.001779
40  Zip or postal code_66216.0    0.001674
21  Zip or postal code_66030.0    0.001381
6                 City_Gardner    0.001236
10                City_Mission    0.001147
26  Zip or postal code_66202.0    0.001118
8                  City_Lenexa    0.000735
25  Zip or postal code_66085.0    0.000450
39  Zip or postal code_66215.0    0.000283
42  Zip or postal code_66219.0    0.000223
19         City_Westwood Hills    0.000146
43  Zip or postal code_66220.0    0.000115
38  Zip or postal code_66214.0    0.000114
35  Zip or postal code_66211.0    0.000053
17            City_Spring Hill    0.000002
41  Zip or postal code_66218.0    0.000000
20  Zip or postal code_66013.0    0.000000
30  Zip or postal code_66206.0    0.000000
47  Zip or postal code_66225.0    0.000000
24  Zip or postal code_66083.0    0.000000

In [ ]:
# Creating new train and test data frames with only the importance features.
train_3 = train_2[impt_feat]
test_3 = test_2[impt_feat]

In [239]:
# Fitting model with train data with all features and testing said model on test data will all features.
model.fit(train_2, train_target_log)
pred_target_log = model.predict(train_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print('Below model was trained and tested with all features')
print('MAE:', mae)
pred_target_log = model.predict(test_2)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE:', mae)


# Fitting model with train data with only the important features and testing said model on test data with only the important features.
model.fit(train_3, train_target_log)
pred_target_log = model.predict(train_3)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(train_target, pred_target))
print()
print('Below model was trained and tested with only features with importance greater than 0')
print('MAE:', mae)
pred_target_log = model.predict(test_3)
pred_target = np.exp(pred_target_log)
mae = np.round(mean_absolute_error(test_target, pred_target))
print('MAE:', mae)

Below model was trained and tested with all features
MAE: 58724.0
MAE: 73305.0

Below model was trained and tested with only features with importance greater than 0
MAE: 60275.0
MAE: 74701.0


Unfortunately, the results were not promising as the MAE for both the train and test data increased. While it would be logical to remove features that have no importance, it wasn't in this case.

## Conclusion

When I first started this project, the main goal was to expand a project I completed in school, and utilize more complex machine learning regressors to create a semi-accurate model to predict the selling prices for houses in Johnson County Kansas. 

Fast forward to the end of this project, I have become discouraged at the results. After initially exploring through six different model possibilities, I honed in on three possible models to further explore: support vector machine regressor, random forest regressor, and gradient boosting regressor. 

In the support vector machine notebook, I enjoyed learning more about the more impactful parameters that played a role, along with making the sure the notebook was well organized and documented properly. When I got to the Random Forest notebook, I realized that this feeling didn't occur anymore, and the quality of the notebook decreased. The main thing I told myself during this moment was that this wasn't the right model, and that the gradient boosting regressor would make everything okay. 

During the gradient boosting regressor notebook, I soon realized that it seemed that the lowest MAE any model could have on the test data was around 65,000-70,000. While not the worst MAE, this would indicate that any model would be off on average by $65000 in determining the selling prices for houses in Johnson County. Additionally, whenever I tinkered with the parameters, it seemed that only the training data score would decrease. When attempts to alleviate over fitting were successful, it came with a cost of the test data MAE growing worse.

During this notebook, I once again somewhat attempted to utilize different parameter options for the Randomized Search CV function in hopes of helping the model generalize better to unseen data, but the model only seemed to grow strong in terms of the training data. While lowering the possible parameter combination seemed to help the model generalize better, it once again came with the cost of the model's test data MAE increasing. 

The last hurdle for this project was attempting to utilize only the features that were important. Unfortunately, this didn't prove a drastic change. 

While I have no doubt there is more I could do to tinker with this project, the fact of the matter is that I feel that there are other projects that I could complete better, and potentially excel in.

Appreciate everyone who has read into any of this at all.
